src/Servant/Client/Client.purs

```diff
@@ -5,114 +5,242 @@ module Servant.Client.Client
-  -> QueryParams params
-  -> Headers headers
-  -> Decoder parsed response
-  -> m response
-buildGetRequest routeP captures params headers decoder =
-  buildRequest_ identity routeP captures params headers decoder
-
-buildPostRequest
-  :: forall m route captures params headers body response parsed parsed'.
-     MonadAsk ClientEnv m
-  => MonadError AjaxError m
-  => MonadAff m
-  => RouteBuilder route captures params headers body response
-  => RouteProxy route
-  -> Captures captures
-  -> body
-  -> QueryParams params
-  -> Headers headers
-  -> Decoder parsed response
-  -> Encoder body parsed'
-  -> m response
-buildPostRequest routeP captures body params headers decoder encoder =
-  buildRequest_ addContent routeP captures params headers decoder
-  where
-  addContent = _{content = pure $ encoder.print $ encoder.encode body}
-
-buildDeleteRequest
-  :: forall m route captures params headers body response parsed parsed'.
-     MonadAsk ClientEnv m
-  => MonadError AjaxError m
-  => MonadAff m
-  => RouteBuilder route captures params headers body response
-  => RouteProxy route
-  -> Captures captures
-  -> body
-  -> QueryParams params
-  -> Headers headers
-  -> Decoder parsed response
-  -> Encoder body parsed'
-  -> m response
-buildDeleteRequest = buildPostRequest
+newtype SuspendedRoute =
+  SuspendedRoute { queryString :: Maybe String
+                 , path :: Array String
+                 , body :: Maybe RequestBody
+                 , headers :: Array (Tuple String String)
+                 }
+
+defaultSuspendedRoute :: SuspendedRoute
+defaultSuspendedRoute = SuspendedRoute { queryString: Nothing
+                                       , path: []
+                                       , body: Nothing
+                                       , headers: []
+                                       }
+
+class RunClient m where
+  -- | How to make a request.
+  runRequest :: forall a. Affjax.Request a -> m (Affjax.Response a)
+
+instance defaultRunClient :: (MonadAff m, MonadError AjaxError m) => RunClient m where
+  runRequest = affjax
+
+class HasClient (r :: Route) (m :: Type -> Type) f | m r -> f where
+  buildClientRoute :: RouteProxy r -> Proxy2 m -> (SuspendedRoute -> f)
+
+instance hasClientPathComponent
+         :: ( HasClient r m f
+            , IsSymbol s
+            )
+         => HasClient (S s :> r) m f where
+  buildClientRoute (RouteProxy :: RouteProxy (S s :> r)) p@(Proxy2 :: Proxy2 m) = \(SuspendedRoute route) ->
+    let pathComponent = reflectSymbol (SProxy :: SProxy s)
+    in buildClientRoute (RouteProxy :: RouteProxy r) p $ SuspendedRoute route { path = cons pathComponent route.path}
+
+else instance hasClientCapture
+         :: ( HasClient r m f
+            , IsSymbol s
+            , ToCapture a
+            )
+         => HasClient (Capture s a :> r) m (Tagged (SProxy s) a -> f) where
+  buildClientRoute (RouteProxy :: RouteProxy (Capture s a :> r)) p@(Proxy2 :: Proxy2 m) = \(SuspendedRoute route) a ->
```

Comment:
Shouldn't you have ` \a (SuspendedRoute route) ->`  instead of ` \(SuspendedRoute route) a ->` as `(Tagged (SProxy s) a -> f)` adds the tagged thing as first argument? (same of other cases)

Reply:
I think it's good the way it is, I don't know how to explain it but let's wait for the tests I guess